[WIP] Sync support

andrey-yantsen opened the pull request:
Hi there! I've started working on Mobile Sync support. If anybody has any comments on the code, or just any feature requests — feel free to tell :)

At the beginning the code can do few useful things:

1. Initialise the "sync" env: tell the plex.tv that the client supports syncing, and the client can be used as sync-target
2. Unfortunately I have to "fake" used platform/device info, because transcoding is depends on this values
3. And finally you can download converted files, by iterating over media parts and looking for ones with required `syncItemId`

The reason why I'm started this is simple: I have a WD My Passport Pro which has the Plex onboard and I'd like to have my "travel" library in sync with the main one. I've already started the project for this (andrey-yantsen/plexiglas) but for now there is nothing to see there.

fix #134

Hellowlol commented:
Hi! Thanks for the PR!

How do you actually sync the items? Can you add a test?

I would expect that there was a sync function inside the client so you could just pass a list of objects and that would be it.

Any info needed for the transcode can be picked up there this way we don’t need to mess with the init method.

andrey-yantsen commented:
Just to clarify: the PR is not finished now, it's an early alpha version :)

> How do you actually sync the items?

The list of "items-to-sync" is controlled in Plex Web, you can add any item to the list for a specific device.

In the following code block I will show the way to download all suitable content (actually I've not yet reached the "downloading" stage in my project, so I'm not sure if this really working now):

```
from plexapi import sync, utils
from plexapi.myplex import MyPlexAccount
sync.init()
plex = MyPlexAccount()

for item in plex.syncItems().items:
    if item.status.state != 'complete':
        continue

    for media in item.getMedia():
        for part in media.iterParts():
            filename = '%s.%s' % (item._prettyfilename(), part.container)
            url = item._server.url(part.key)
            filepath = utils.download(url, token=plex.authenticationToken, filename=filename, savepath=os.getcwd(),
                                      session=item.server()._session, showstatus=True)

```

> Can you add a test?

Yep, of course, on some latter stage :)

> I would expect that there was a sync function inside the client so you could just pass a list of objects and that would be it.
> Any info needed for the transcode can be picked up there this way we don’t need to mess with the init method.

As far as I can see the only way for Plex to detect transcoding options, is thru target platform. Here is what I see in the Plex logs with real platform info:

```
ERROR - Unable to find client profile for device; platform=Darwin, platformVersion=17.7.0, device=
Darwin, model=
```
And the process goes smoothly when I mimic my iPhone SE.

Also here is the request which should be sent to add an item to sync list:

```
POST https://plex.tv/devices/779755E8-54E2-4D0F-A7D9-E2FF8171CC55/sync_items?X-Plex-Device-Name=Yantsen%20-%20iPhone%20An HTTP/1.1

Content-Type:	application/x-www-form-urlencoded

SyncItem[Location][uri]:        library://ac0a6a57-522b-4dcf-9cb8-0a58b16c6726/item/%2Flibrary%2Fmetadata%2F2428
SyncItem[MediaSettings][audioBoost]:100
SyncItem[MediaSettings][maxVideoBitrate]:4000
SyncItem[MediaSettings][musicBitrate]:192
SyncItem[MediaSettings][photoQuality]:74
SyncItem[MediaSettings][photoResolution]:1920x1080
SyncItem[MediaSettings][subtitleSize]:100
SyncItem[MediaSettings][videoQuality]:100
SyncItem[MediaSettings][videoResolution]:1280x720
SyncItem[Policy][scope]:        all
SyncItem[Policy][unwatched]:    0
SyncItem[Policy][value]:        0
SyncItem[contentType]:          video
SyncItem[machineIdentifier]:    607c8c938b50eef734456f8b9da94b5d02339ce5
SyncItem[metadataType]:         movie
SyncItem[rootTitle]:            Test
SyncItem[title]:                Test1
```

As you can see there are nothing here about container and/or supported codecs list, only basic/common transcoding stuff.

Hellowlol commented:
You dont really need to do anything to make the sync happens. Seems you just upload zhe shit to a queue, and the client handles the rest from there.

Anyway to send a sync command should be as easy as:

inside client.py
````

def sync_item(self, item): 
    #magic happens here.
    # Maybe grab stuff that you need from the client like platform etc
    # name, title from item etc.

   
````

Looking forward to test the pr :)

andrey-yantsen commented:
> I dont think syncing to computers is supported by plex at all.

It's not supported out-of-the-box, but we're the engineers, and we can force software to do what we want! :)

> Do you just want to copy things from another shared plex server to your own?

I want to copy optimised copies of things from any plex server to an external hdd

Hellowlol commented:
Oh, thats easy.  Its pretty much supported out of the box.

See the docs for.

https://python-plexapi.readthedocs.io/en/latest/modules/video.html?highlight=getStreamURL%20#plexapi.video.Movie.download

note the comment about **kwargs – Additional options passed into getStreamURL().
You can use this to transcode.

andrey-yantsen commented:
Yep, I saw this option, I plan to support it in my project, but it's not on my priority list...

`getSteamURL()` not that powerful, easy and cute like Mobile Sync: e.g. with Sync you can mark to sync "3 unwatched episodes", "3 unwatched movies with highest rating", "3 unwatched movies released recently", and so on. And all those media would be automatically rotated while you marking something as watched, when new releases become available.

Hellowlol commented:
I think im missing something here. Do you expect to do this from the plex ui?
If would be easy to get movies/episodes based on some condition run them in some cron script.

andrey-yantsen commented:
> Do you expect to do this from the plex ui?
You're exactly right :) https://support.plex.tv/articles/201082477-quick-guide-to-mobile-sync/

> If would be easy to get movies/episodes based on some condition run them in some cron script.
Man... I just like the Mobile Sync approach, where you don't have to code the filters by yourself, but you can easily choose the required things with UI and this UI doesn't require any dev-knowledge at all, even my wife would be able to use it! :D

Hellowlol commented:
Oh, i think i get it now. You faking so plexapi is detected as a client in the plex webui... imo that should be done with your own project!

Nice idea btw!

mikes-nasuni commented on plexapi/base.py:
```diff
@@ -144,7 +144,8 @@ def fetchItems(self, ekey, cls=None, **kwargs):
             on how this is used.
         """
-        data = self._server.query(ekey)
+        timeout = kwargs.pop('timeout', None)
```
This should probably be added as a real keyword arg timeout=None in the function definition and not pulled out of kwargs.

mikes-nasuni commented on plexapi/myplex.py:
```diff
@@ -707,7 +707,7 @@ def delete(self):
 
 
-def _connect(cls, url, token, timeout, results, i):
+def _connect(cls, url, token, timeout, fast, results, i, connected_event=None):
```
If X_PLEX_ENABLE_FAST_CONNECT is a global configuration option. Should we instead always use fast=True or false based on that, rather than passing it in as an argument?

mikes-nasuni commented on plexapi/utils.py:
```diff
@@ -152,15 +152,18 @@ def threaded(callback, listargs):
-    for thread in threads:
-        thread.join()
-    return results
+    while not connected_event.is_set():
```
Just curious, what was the issue that led to this change?

andrey-yantsen replied:
Most times I run the my things outside my home network, when the local IP of the server is not available, and it takes annoyingly long time to establish the connection

mikes-nasuni replied:
Makes sense, thanks.

andrey-yantsen commented:
@mikes-nasuni thank you for the fast reaction, I've made the changes which you've suggested :)

Hellowlol commented on plexapi/utils.py:
```diff
@@ -152,15 +152,18 @@ def threaded(callback, listargs):
+        if all([not t.is_alive() for t in threads]):
+            break
+        time.sleep(0.05)
+    return list(filter(lambda r: r is not None, results))
```
Use a list comprehension

Hellowlol commented on plexapi/sync.py:
```diff
@@ -1,42 +1,103 @@
-        url = '/sync/%s/%s/files/%s/downloaded' % (
-            self._device.clientIdentifier, server.machineIdentifier, sync_id)
-        server.query(url, method=requests.put)
+    def markDownloaded(self, media: Playable):
```
This need to be support python 2.

Hellowlol commented on plexapi/sync.py:
```diff
@@ -1,42 +1,103 @@
     def server(self):
-        server = list(filter(lambda x: x.machineIdentifier == self.machineIdentifier, self._servers))
+        server = list(filter(lambda x: x.clientIdentifier == self.machineIdentifier, self._server.resources()))
         if 0 == len(server):
```
List comp please

Hellowlol commented on plexapi/sync.py:
```diff
@@ -1,42 +1,103 @@
         return server[0]
 
-    def getMedia(self):
+    def getMedia(self, timeout=None):
```
What up with the timeout? Why not just change the plex.Timeout?

andrey-yantsen replied:
Not all operations require 60s+ timeout, and wrapping `getMedia()` calls with changing timeout back-and-forth would look strange.
You think that just globally change `TIMEOUT` in my code would be more than enough and wouldn't cause any strange problems?

mikes-nasuni replied:
> You think that just globally change TIMEOUT in my code would be more than enough and wouldn't cause any strange problems?

It shouldn't cause any problems. timeout is essentially just telling the requests module when to give up if not getting a response. It's really only useful to change to poor or slow network issues. Having different timeouts for different calls in to the API seems incorrect unless you have a really strong case for it.

andrey-yantsen commented:
@Hellowlol thank you for the review! I've fixed most of the issues, but I'm not sure on `timeout` arg in `getMedia()` / `fetchItems()`.

mikes-nasuni commented on plexapi/myplex.py:
```diff
@@ -698,7 +707,7 @@ def delete(self):
 
 
-def _connect(cls, url, token, timeout, results, i):
+def _connect(cls, url, token, timeout, results, i, connected_event=None):
```
What is connected_event used for, can you add a comment in the doc string here about what what is is?

mikes-nasuni commented on plexapi/__init__.py:
```diff
@@ -17,6 +17,7 @@
 VERSION = '3.0.6'
 TIMEOUT = CONFIG.get('plexapi.timeout', 30, int)
 X_PLEX_CONTAINER_SIZE = CONFIG.get('plexapi.container_size', 100, int)
+X_PLEX_ENABLE_FAST_CONNECT = CONFIG.get('plexapi.enable_fast_connect', False, bool)
```
Can you please update the configuration.rst file with what X_PLEX_ENABLE_FAST_CONNECT is and when we would want to use it.

https://github.com/pkkid/python-plexapi/blob/master/docs/configuration.rst

mikes-nasuni commented on plexapi/sync.py:
```diff
@@ -1,31 +1,58 @@
+        self.clientIdentifier = clientIdentifier
 
     def _loadData(self, data):
         self._data = data
```
There should be a documentation header on head class in this file explaining all the attributes and why they exist. A good example format is the video.py file.

https://github.com/pkkid/python-plexapi/blob/master/plexapi/video.py

mikes-nasuni commented on plexapi/sync.py:
```diff
@@ -35,8 +62,53 @@ def getMedia(self):
+        media._server.query(url, method=requests.put)
+
+
+class SyncList(PlexObject):
```
Also needs documentation.

mikes-nasuni commented:
This is starting to shape up nicely, thanks.

andrey-yantsen commented on plexapi/config.py:
```diff
@@ -60,4 +60,5 @@ def reset_base_headers():
         'X-Plex-Device': plexapi.X_PLEX_DEVICE,
         'X-Plex-Device-Name': plexapi.X_PLEX_DEVICE_NAME,
         'X-Plex-Client-Identifier': plexapi.X_PLEX_IDENTIFIER,
+        'X-Plex-Sync-Version': '2',
```
I've added this because it doesn't require anything from the client, it just reports to the server the we can work with Sync. Basically without this header you can't add new SyncItems.

andrey-yantsen commented:
Hey guys, please review :) Looks like I almost finished with the development (actually I already have my app syncing videos for me for almost a week now), the one huge thing which is left — is tests... And make all of this work with audio/photos, but it shouldn't be that hard at current stage. I hope I will be able to finish them during next week.

Hellowlol commented:
Sorry, I’ll review this later today. Been busy.

andrey-yantsen commented:
FYI there is a typo in `__init__.py`:

```
X_PLEX_PLATFORM = CONFIG.get('header.platorm', uname()[0])
```

`platorm` instead of `platform`. Probably we should fix this, but this would be a breaking change :(

andrey-yantsen commented:
Ok. At the moment I have tests for video/show/season/episode adding to synclist and reading from it. Here is what left to do:

* Test downloading
* Add method sync() for audio, photo & playlist

Anything else?

Hellowlol commented on plexapi/library.py:
```diff
@@ -543,6 +543,53 @@ def _cleanSearchSort(self, sort):
+    def sync(self, client, policy, media_settings, title=None, sort=None, libtype=None, **kwargs):
+        """ Add current library section as sync item for specified device.
+            See description of :func:`plexapi.library.LibraySection.search()` for details about filtering / sorting.
+
```
Example is good but can you add docs string for args?

andrey-yantsen replied:
Yep, sure, I'm working on tests now, and fixing all the docs :)

Hellowlol commented on plexapi/myplex.py:
```diff
@@ -378,6 +379,57 @@ def optOut(self, playback=None, library=None):
+        """ Returns an instance of :class:`~plexapi.sync.SyncItems` for specified client
+
+            Arguments:
+                 clientId (str): an identifier of a client for which you need to get SyncItems. Would be set to current
```
allow for both a client and a clientid as a string.

Hellowlol commented on plexapi/myplex.py:
```diff
@@ -378,6 +379,57 @@ def optOut(self, playback=None, library=None):
+        """ Adds specified sync item for the client
+
+            Arguments:
+                 client (:class:`~plexapi.myplex.MyPlexDevice`): pass
```
doc what this client, sync_item is used for.

Hellowlol commented on plexapi/server.py:
```diff
@@ -422,6 +422,16 @@ def url(self, key, includeToken=None):
         return '%s%s' % (self._baseurl, key)
 
+    def refreshSynclist(self):
+        return self.query('/sync/refreshSynclists', self._session.put)
```
add missing doc string

Hellowlol commented on plexapi/server.py:
```diff
@@ -422,6 +422,16 @@ def url(self, key, includeToken=None):
+        return self.query('/sync/refreshSynclists', self._session.put)
+
+    def refreshContent(self):
+        return self.query('/sync/refreshContent', self._session.put)
```
add missing doc string

Hellowlol commented on plexapi/sync.py:
```diff
@@ -1,42 +1,256 @@
+        status (:class:`~plexapi.sync.Status`): current status of the sync
+        mediaSettings (:class:`~plexapi.sync.MediaSettings`): media transcoding settings used for the item
+        policy (:class:`~plexapi.sync.Policy`): the policy of which media to sync
+        location (str): unknown
```
self.clientidentifier is not listed as a attribute.

Hellowlol commented on plexapi/sync.py:
```diff
@@ -1,42 +1,256 @@
+        """ Returns :class:`~plexapi.myplex.MyPlexResource` with server of current item.
+        """
+        server = [s for s in self._server.resources() if s.clientIdentifier == self.machineIdentifier]
         if 0 == len(server):
```
can you do len(server) == 0 instead,

Hellowlol commented on plexapi/sync.py:
```diff
@@ -1,42 +1,256 @@
+        media._server.query(url, method=requests.put)
+
+    def delete(self):
+        url = SyncList.key.format(clientId=self.clientIdentifier)
```
Missing doc string

Hellowlol commented on plexapi/sync.py:
```diff
@@ -1,42 +1,256 @@
+        self.clientId = data.attrib.get('clientIdentifier')
+        self.items = []
+
+        for elem in data:
```
Can we use .iter('SyncItem') here?

Hellowlol commented on plexapi/video.py:
```diff
@@ -77,6 +78,31 @@ def markUnwatched(self):
+        return self._prettyfilename()
+
+    def sync(self, client, policy, media_settings, title=None):
+        """ Add current video as sync item for specified device.
```
add docs for args, and keywork arguments.

Hellowlol commented on tests/test_sync.py:
```diff
@@ -0,0 +1,82 @@
+MAX_RETRIES = 3
+
+
+def get_new_synclist(account, client_id, old_sync_items=[]):
```
old_sync_items is mutable, is this intended?

andrey-yantsen replied:
I've already rewrote this part locally, anyway :D

andrey-yantsen commented:
Heyhey, @Hellowlol, thank you for the review! I've pushed another pile of code, where I've done few things:

1. Fixed part of your review request
2. Marked all (I hope that I marked them all) as `TODO`, to not miss it later
3. Simplified initialisation of `sync.MediaSettings` in order to get this working easily with other content types
3. Added `sync()` method to `audio`, `photo` & `playlists`
4. Each sync-related method now can accept `client` as an object, or `clientId` as a string
5. Covered this stuff with creepy tests

In order to get sync/tests for photos working PR #292 need to be applied.

Hellowlol commented:
Let me know when this is ready, i want to squash and merge this before i do the small prs.

andrey-yantsen commented:
Yep, sure. I hope to finish this by tomorrow.

Hellowlol commented:
Thanks for the doc changes. I havnt really looked at the code before now. There is a lot of repetitions regarding the sync methods. Anyway we could simplify this?

andrey-yantsen commented:
Uh... From the top of my head — I can add a mixin `Syncable` with some basic implementation of the method, but, IMO, because of different content types (video, audio, photo) require different options, and it's possible to sync so many different, not directly related in OOP to each other, stuff (library, playlist, movie, tv-show episode, etc.) there are still be some repetition... 

@Hellowlol Do you have any better ideas?

Hellowlol commented:
Using a mixin was my tought too be if there are too many options we might just leave it as it is. Anyway are you done?

andrey-yantsen commented:
Yep, it's done I think. Probably I'll find something else later, but for now I have nothing to add :)

Hellowlol commented:
Awesome, thanks!

Hellowlol commented:
Can you fix the inline imports so the style is consistent? We don't seem to use any other relative imports,

andrey-yantsen commented:
Yep, sure, I've just pushed it
